chat.js

```diff
@@ -920,25 +925,53 @@ Chat.uncacheTree = function (root) {
 	} while (uncache.length > 0);
 };
 
-Chat.loadCommands = function () {
-	if (Chat.commands) return;
+Chat.loadPlugins = function () {
+	if (Chat.commands || Chat.filter) return;
 
 	FS('package.json').readTextIfExists().then(data => {
 		if (data) Chat.package = JSON.parse(data);
 	});
 
 	let baseCommands = Chat.baseCommands = require('./chat-commands').commands;
 	let commands = Chat.commands = Object.assign({}, baseCommands);
+	let chatfilters = Chat.filters = [];
 
-	// Install plug-in commands
+	// Install plug-in commands and chat filter
 
 	// info always goes first so other plugins can shadow it
 	Object.assign(commands, require('./chat-plugins/info').commands);
 
 	for (let file of FS('chat-plugins/').readdirSync()) {
 		if (file.substr(-3) !== '.js' || file === 'info.js') continue;
-		Object.assign(commands, require('./chat-plugins/' + file).commands);
+		const plugin = require(`./chat-plugins/${file}`);
+
+		Object.assign(commands, plugin.commands);
+
+		const filter = plugin.chatfilter;
+		if (filter) {
+			if (typeof filter !== 'function') {
+				require('./crashlogger')(`This chatfilter is not a function`, `Loading a chatfilter`, {
+					file: `File location: ../chat-plugins/${file}`,
+				});
+			} else { // only use chatfilters that are a function
+				chatfilters.push(filter);
+			}
+		}
 	}
+
+	Chat.filter = function (message, user, room, connection, targetUser) {
+		if (!chatfilters.length) return message;
+
+		let outputs = [];
+		for (let i = 0; i < chatfilters.length; i++) {
+			const output = chatfilters[i].call(this, message, user, room, connection, targetUser);
+			outputs.push(output);
+		}
+
+		// If any filters returned false, then we do not want to send their
+		// messages to the room, otherwise, we will send their message
+		if (!outputs.includes(false)) return message;
+	};
```

Comment:
o_O Allowing for message rewrites via the chat filter is essential...

> Also, the output of each previous filter step should be passed to the next as the main argument.
The original message should be passed as well, as another argument.